mbedTLS: clean-up insecure/deprecated code, and other fixes

wyattoday opened the pull request:
We use mbedTLS quite a bit, and I've been meaning to do a comb through the code used in curl for a while now. This PR makes quite a few changes but none of them are particularly revolutionary in themselves (and don't really warrant separate PRs for each little change).

But I will break down the changes here so it's more clear what I've done and why. Long story short: this increases the security and reduces some cruft:


1. With `MBEDTLS_SSL_PROTO_TLS1_2` not enabled, the mbedTLS code was not able to connect to any server due to broken logic in curl's `mbed_set_ssl_version_min_max()`. Now it correctly sets the minimum supported TLS version based on what is compiled in the library.

2. Fix `MBEDTLS_DEBUG_C` logging support (namely, if the library is built with debugging enabled, then enable debugging in here). Don't require the user to uncomment code, etc. This simplifies debugging-logs into one configuration (rather than 2 very similarly named but broken if one is enabled but the other isn't).

3. If debugging is enabled, move the debugging enabling earlier in the `mbed_connect_step1()` so that verbose errors are actually displayed if failures happen (see the first point -- it would've made debugging that issue easier).

4. Remove mbedTLS 3.0.0-specific bug fixes. That's a deprecated and insecure version (people on 3.0 through 3.5 should upgrade to the latest 3.x branch 3.6.x as I write this, which is also an LTS release). Keeping this noise around makes it very hard to audit the code correctly (because who is actually testing the deprecated 3.0-3.5 releases? Not us, that's for sure.)

5. Remove mbedTLS 2.x and older support. It no longer receives security fixes (or any bug fixes) and should be considered harmful to use. Use mbedTLS latest LTS release (3.6.x). [From mbedTLS](https://github.com/Mbed-TLS/mbedtls/releases/tag/mbedtls-2.28.10):

> Mbed TLS 2.28.10 is the last release of the 2.28 LTS and won't receive bug fixes or security fixes anymore.
Users are advised to upgrade to a maintained version.

6. Remove the constant `mbedtls_x509_crt_profile_fr` and instead use mbedTLS-included profile `mbedtls_x509_crt_profile_next` with `mbedtls_ssl_conf_cert_profile()`.
    This will follow the latest standards as new mbedTLS versions are released (rather than being stuck-in-time until someone comes along to fix what was hard-coded here). This has the immediate benefit of no longer supporting SHA1 certs and insecure RSA key-lengths (1024). This fix immediately prevents previously possible MITM attacks (SHA1 hashes and RSA-1024 keys can be forged relatively easily by nation-state actors and criminal organizations with deep-pockets).

7. Removed `MBEDTLS_HAVEGE_C` entropy source (a potential security vulnerability with low-entropy source). Or, [straight from the horse's mouth](https://mbed-tls.readthedocs.io/en/latest/kb/how-to/add-entropy-sources-to-entropy-pool/):

> **Note:** The HAVEGE random generator is considered reasonable but not good. Please do not base your full entropy on this.

And curl, was of course, basing its full entropy on this. So, just remove this foot-gun.


8. Added [predictive resistance](https://mbed-tls.readthedocs.io/en/latest/kb/how-to/add-a-random-generator/#enabling-prediction-resistance) to the random number generator (adding more entropy to the RNG).

9. Split the random number generator into initialization, the actual random generation, and the "freeing" of the resources. This significantly reduces the overhead of using the RNG.

10. Removed the separate RNG function in the TLS connect stage (instead use the "main" one) and remove the ad-hoc threading support. Instead properly document how to enable threading in mbedTLS. As it was, other internals of mbedTLS could have race conditions (in the RSA module in particular) if `MBEDTLS_THREADING_C` was *not* enabled. And if it is enabled, then these race-conditions cannot happen. And also, if MBEDTLS_THREADING_C is enabled then the RNG functions [are fully thread-safe](https://mbed-tls.readthedocs.io/en/latest/kb/development/thread-safety-and-multi-threading/).

   So, the previous ad-hoc threading support was both partial and broken.


11. Enable support for disabling `MBEDTLS_PEM_PARSE_C`.

12. Add support for `CURLOPT_SSLCERTTYPE` so user can specify `PEM` or `DER` and get faster execution.

~~13. Add `CURL_DISABLE_CERTINFO` and `CURL_DISABLE_CIPHER_SELECT` build options. This is part of an upcoming larger PR to reduce unnecessary code (for end-users who don't need these levers). This is especially important on embedded CPUs (another upcoming PR) where every bit of flash, ram, and CPU is precious. "Cert info" is nice to have if you're debugging, but in my opinion shouldn't be enabled by default in the library. Cipher select is a relic from the past now that TLS 1.3 doesn't let you select insecure ciphers (and, for mbedTLS in particular, you can actively not-compile any algorithm you consider to be too weak).~~ Coming in a followup PR.

vszakats commented on lib/vtls/cipher_suite.c:
```diff
@@ -23,7 +23,8 @@
 #include "../curl_setup.h"
 
-#if defined(USE_MBEDTLS) || defined(USE_RUSTLS)
+#if !defined(CURL_DISABLE_CERTINFO) && \
+     (defined(USE_MBEDTLS) || defined(USE_RUSTLS))
```
```suggestion
#if !defined(CURL_DISABLE_CERTINFO) && \
  (defined(USE_MBEDTLS) || defined(USE_RUSTLS))
```

vszakats commented on lib/vtls/mbedtls.c:
```diff
@@ -558,7 +502,7 @@ static int mbed_verify_cb(void *ptr, mbedtls_x509_crt *crt,
 
   if(*flags) {
-#if MBEDTLS_VERSION_NUMBER < 0x03000000 || !defined(MBEDTLS_X509_REMOVE_INFO)
+#if !defined(MBEDTLS_X509_REMOVE_INFO)
```
```suggestion
#ifndef MBEDTLS_X509_REMOVE_INFO
```

vszakats commented on lib/vtls/mbedtls.c:
```diff
@@ -1427,29 +1395,24 @@ static CURLcode mbedtls_random(struct Curl_easy *data,
 
-  if(!ret)
-    ret = mbedtls_ctr_drbg_random(&ctr_drbg, entropy, length);
+#if defined(MBEDTLS_CTR_DRBG_C)
```
```suggestion
#ifdef MBEDTLS_CTR_DRBG_C
```

vszakats commented on lib/vtls/mbedtls.c:
```diff
@@ -1516,34 +1479,78 @@ static CURLcode mbedtls_connect(struct Curl_cfilter *cf,
+  mbedtls_entropy_init(&rng.entropy);
+
+
+#if defined(MBEDTLS_CTR_DRBG_C)
+  int ret = mbedtls_ctr_drbg_seed(&rng.drbg,
+      mbedtls_entropy_func, &rng.entropy,
+      NULL, 0);
+#elif defined(MBEDTLS_HMAC_DRBG_C)
+#if defined(MBEDTLS_SHA256_C)
```
```suggestion
#ifdef MBEDTLS_CTR_DRBG_C
  int ret = mbedtls_ctr_drbg_seed(&rng.drbg,
      mbedtls_entropy_func, &rng.entropy,
      NULL, 0);
#elif defined(MBEDTLS_HMAC_DRBG_C)
#ifdef MBEDTLS_SHA256_C
```

vszakats commented on lib/vtls/mbedtls.c:
```diff
@@ -1516,34 +1479,78 @@ static CURLcode mbedtls_connect(struct Curl_cfilter *cf,
+     Only use this if you have ample supply of good entropy.*/
+  mbedtls_ctr_drbg_set_prediction_resistance(&rng.drbg,
+                                             MBEDTLS_CTR_DRBG_PR_ON);
+
```
```suggestion
```

bagder commented:
> Add CURL_DISABLE_CERTINFO and CURL_DISABLE_CIPHER_SELECT build options. This is part of an upcoming larger PR

I propose you don't use them in this PR. We have tests that verify that `CURL_DISABLE_` macros are documented and possible to set with configure, so adding them partially causes some friction.

wyattoday commented:
All tests are now passing except:

`CM mbedtls-pkg MultiSSL`: uses mbedTLS 2.8.x which is no longer supported by mbedTLS and it's explicitly dropped by this PR.

`CM mbedtls-pkg !pc`: Also uses mbedTLS 2.8.x which is no longer supported by mbedTLS and it's explicitly dropped by this PR.

`AM mbedtls clang`: This one is odd. It might be an issue of mbedTLS not being built with threading enabled, and the tests are run on threads. This one is hard to parse, but I'm not finding these errors in real-life.




Also, the workaround to the `random()` being called before the `init()` function was to keep track of the state of the TLS library inside the mbedtls.c file. Which is ... a solution. But I'm not sure if this is just a bug in curl (did you intend for this behavior to happen -- TLS libraries called before init()?). If it is a bug I'll make a PR that undoes this workaround.

If the behavior is working as intended, then this workaround is the best way to do things.

Other than what I pointed out above (and any other notes from @badger or anyone else) then everything should be good to merge.

wyattoday commented:
@bagder 

OK. I've added an `#error ` early in the source to tell users exactly why the build is failing (for old mbedTLS versions).

I've also cleaned up the `mbed_debug()` function to bail out on "empty" data (this seems like it was causing valgrind to get angry). `mbed_debug` is one function I didn't touch, but was newly "exposed to the world" because I removed cruft surrounding how debugging is enabled ([see point 2 above](https://github.com/curl/curl/pull/18161#issue-3289326228)):

> Fix `MBEDTLS_DEBUG_C` logging support (namely, if the library is built with debugging enabled, then enable debugging in here). Don't require the user to uncomment code, etc. This simplifies debugging-logs into one configuration (rather than 2 very similarly named but broken if one is enabled but the other isn't).

If debugging continues to cause problems with valgrind, disabled debugging in the mbedTLS build itself is the way to get things "back to the way they were before" (i.e. not outputting debugging information in the CI builds). That is, build mbedTLS with the `MBEDTLS_DEBUG_C` commented out in `mbedtls_config.h`

wyattoday commented:
OK. I modified the `linux.yml` CI script to correctly configure the mbedTLS build to unset `MBEDTLS_DEBUG_C` (because it was breaking the tests that didn't expect a whole bunch of extra junk to be output.

I also commented out the `mbedtls-pkg MultiSSL` and `mbedtls-pkg !pc` tests because they're running on versions of Linux that have very old "`libmbedtls-dev`" packages (2.8.x are not secure, see above PR description). So, testing against broken packages will result in broken builds. I didn't delete them because I didn't know if y'all wanted to upgrade the "base system" to something up-to-date (like "sid" Debian if you want to stay in the Debian/Ubuntu ecosystem).

You might want to clear the cache for the mbedTLS builds. Or not. Not quite sure if they're cached repository wide, PR wide, etc.

wyattoday commented:
Yeah, OK, they're cached repository-wide.

Clear the mbedTLS CI cache and re-run the tests and they pass (I tested it already by commenting out the cache-lookup and it passed). I put back the cache-lookup but you'll need to clear the cache for it to build mbedTLS correctly.

So, code is fixed and CI is fixed to support the more robust code.

If there are no other comments, suggestions, this feels ready-to-merge to me.

icing commented:
> OK. I modified the `linux.yml` CI script to correctly configure the mbedTLS build to unset `MBEDTLS_DEBUG_C` (because it was breaking the tests that didn't expect a whole bunch of extra junk to be output.
> 
> I also commented out the `mbedtls-pkg MultiSSL` and `mbedtls-pkg !pc` tests because they're running on versions of Linux that have very old "`libmbedtls-dev`" packages (2.8.x are not secure, see above PR description). So, testing against broken packages will result in broken builds. I didn't delete them because I didn't know if y'all wanted to upgrade the "base system" to something up-to-date (like "sid" Debian if you want to stay in the Debian/Ubuntu ecosystem).
> 
> You might want to clear the cache for the mbedTLS builds. Or not. Not quite sure if they're cached repository wide, PR wide, etc.

I think I'd prefer the super-verbose debug logging to be off by default, not as a special configuration used for CI. As a maintainer, I need to build mbedtls versions to test other things. And just using a curl debug build should not prevent that.

wyattoday commented:
> And just using a curl debug build should not prevent that.

The debug configuration I changed in the CI was disabling debugging in mbedTLS (i.e. commenting out the `MBEDTLS_DEBUG_C` variable in `mbedtls_config.h`). Which is a default "production" value for mbedTLS. It doesn't change whether debugging is enabled in curl or not.

vszakats commented:
I take issue with adding new features and dropping support for
a major backend version within the same commit. It makes it
difficult to track what belongs to which change and they are also
different discussion and decisions.

Would it be possible to add the new features without removing
support for 2.x? Then decide about the latter separately?

mbedTLS 2.x is still shipping with all stable Debian versions:
https://sources.debian.org/src/mbedtls/

This patch also removes CI jobs for 2.x. These jobs were testing
more than just mbedTLS 2.x features, which would need to be
moved to other jobs that could take them.

Dropping 3.x minor version specific bugfixes seems to be less
of a practical issue (?), but these are also probably better handled
in their own commit, to see what's gained by doing this. If it's
a couple of lines, maybe it'd be worth keeping it, like done in
other cases that I can remember, to not cause build troubles 
downstream.

icing commented:
> Regarding a new define to enable debugging -- I personally think that's bad user-design. If a user builds mbedTLS with debugging then it should use debugging. 

I do not agree. Turns out, the mbedTLS I built locally has for years has debugging enabled, but I have never wanted to debug it. What I *have* done several times is debug how curl works in a build with mbedTLS.

Now, you could say I was building it wrong, but it seems to be an mbedTLS thing by default (I did not do anything extra). So, you might be telling all curl + mbedTLS users in the world that they are wrong.💁🏻‍♂️

In my experience, people who debug curl usually want to debug curl and not the libraries it is linked with, even though they might offer debug hooks.

So much to that point.

Your other point, that it is not always easy to land a PR to curl as the project is quite active, I sympathise with. 

The particular problem with this PR is that it *removes* support for older versions. That is something we are very careful about in this project and require good reasons for. If the PR covered only enhancements, passing our test suites (which is an absolute must), we could take it in much easier and add what things we might feel missing ourselves.

So, it's a bit of misfortune here, I'd say. But nothing's really lost. Let's revisit this when you have more time. It does not run away.

bagder commented:
Old now with huge merge conflicts. If this comes back, it is better to make new smaller PRes.
